Implement multi-schema support

philipp-kleber-avelios opened the pull request:
These changes are an initial experiment to resolve issue #248, by allowing multiple schemas during the snapshot.
I added a new test `HibernateMultiSchemaTest.java`, which currently only checks the exit code of such a multi-schema `diffChangelog` command.

If somebody could review the code and maybe give their input if the changes are sufficient or if additional ones are required, that would be great, thanks!.

philipp-kleber-avelios committed:
Implement multi-schema support

philipp-kleber-avelios committed:
Implement multi-schema support

MalloD12 commented on src/main/java/liquibase/ext/hibernate/database/HibernateDatabase.java:
```diff
@@ -36,13 +36,11 @@ public abstract class HibernateDatabase extends AbstractJdbcDatabase {
 
     private boolean indexesForForeignKeys = false;
-    public static final String DEFAULT_SCHEMA = "HIBERNATE";
+    public static final String DEFAULT_DEFAULT_CATALOG_NAME = "HIBERNATE";
```
Is there any reason why these constants are named `DEFAULT_DEFAUL..`?

philipp-kleber-avelios replied:
I think my idea was that it is the default value for `getDefaultSchemaName`, so that's why there's two defaults. But I agree it's confusing, so I updated it.

MalloD12 reviewed (approved):
**Approved**.

Besides the minor comment I posted this looks good to me. Thank you, @philipp-kleber-avelios for contributing with this PR.

Daniel.

philipp-kleber-avelios committed:
Improve constant names in HibernateDatabase

filipelautert commented on src/main/java/liquibase/ext/hibernate/database/HibernateDatabase.java:
Why `getConnectionCatalogName()` and `getConnectionSchemaName()` are being locked here to constants?

philipp-kleber-avelios replied:
For `getConnectionCatalogName()`, the behavior is the same as before just without one less indirection (it used to call `getDefaultCatalogName()` internally).

For `getConnectionSchemaName()`, the behavior is slightly different from before: 
It used to call `getDefaultSchemaName()` internally, but we can't do this anymore: The new  `getDefaultSchemaName()` implementation calls `super.getDefaultSchemaName()` and this then calls `getConnectionSchemaName()`, so we'd have a call-cycle.

filipelautert commented on src/main/java/liquibase/ext/hibernate/snapshot/SequenceSnapshotGenerator.java:
```diff
@@ -34,6 +34,12 @@ protected void addTo(DatabaseObject foundObject, DatabaseSnapshot snapshot) thro
             Schema schema = (Schema) foundObject;
             HibernateDatabase database = (HibernateDatabase) snapshot.getDatabase();
             for (org.hibernate.boot.model.relational.Namespace namespace : database.getMetadata().getDatabase().getNamespaces()) {
+                boolean namespaceMatchesSchema = (namespace.getName().getSchema() != null && namespace.getName().getSchema().matches(foundObject.getName()))
```
This check is duplicated in `SequenceSnapshotGenerator` and `TableSnapshotGenerator`. Could you extract it to a generic method in `HibernateSnapshotGenerator` , or even move the check there so it's not repeatable? As I believe it would be necessary to `IndexSnapshotGenerator` and other `SnapshotGenerators` too.. or am I overreacting?

philipp-kleber-avelios replied:
Sure, I can extract the check. I don't think we need to move the check though, since from my understanding, `DatabaseObjects` other than `Schema`, `Table` and `Sequence` are added only from the corresponding snapshot-generators. 
So I think if we have the check in those three generators, we don't need the check in the other ones, since the database objects that get to them always already have the correct schema. 
I might be wrong though, I don't fully understand the snapshotting mechanism.

philipp-kleber-avelios replied:
done

philipp-kleber-avelios committed:
Extract schema matching logic to HibernateSnapshotGenerator.

filipelautert reviewed (approved):
Thanks for the improvement @philipp-kleber-avelios !
@tati-qalified if you need any ideas on how to test it please let me know.

tati-qalified commented on src/main/java/liquibase/ext/hibernate/snapshot/SequenceSnapshotGenerator.java:
```diff
@@ -34,6 +34,10 @@ protected void addTo(DatabaseObject foundObject, DatabaseSnapshot snapshot) thro
             HibernateDatabase database = (HibernateDatabase) snapshot.getDatabase();
             for (org.hibernate.boot.model.relational.Namespace namespace : database.getMetadata().getDatabase().getNamespaces()) {
+                if (!schemaMatchesNamespace(schema, namespace)) {
+                    continue;
```
@philipp-kleber-avelios in this piece of code it's checking if the schema doesn't match the namespace, but then it does nothing either way. Did you mean to add something else inside the if block? If not, it should be removed

tati-qalified commented on src/main/java/liquibase/ext/hibernate/snapshot/TableSnapshotGenerator.java:
```diff
@@ -38,13 +41,19 @@ protected DatabaseObject snapshotObject(DatabaseObject example, DatabaseSnapshot
-        Table table = new Table().setName(hibernateTable.getName());
-        Scope.getCurrentScope().getLog(getClass()).info("Found table " + table.getName());
+        if (!schemaMatchesTable(example, hibernateTable)) {
+            Scope.getCurrentScope().getLog(getClass()).info("Skipping table " + hibernateTable.getName() + " for schema " + example.getSchema().getName() + ", because it is part of another one.");
```
The last part `because it is part of another one` isn't too clear, I'd change it to `because it belongs to a different schema.`

tati-qalified reviewed (changes requested):
Thank you @philipp-kleber-avelios for your contribution!
I was wondering if the entities for the test could be narrowed down, I don't think it's necessary to create 13 classes just for one test case.

MalloD12 committed:
- Update method signatures in HibernateDatabase.
- Null check and message updated for TableSnapshotGenerator.
- Rename HibernateMultiSchemaTest function, added comments to make test more reable and added additional assertions.
- Remove a couple of entities to simplify testing.

MalloD12 committed:
Merge branch 'multi-schema-support' of https://github.com/philipp-kleber-avelios/liquibase-hibernate into philipp-kleber-avelios-multi-schema-support

MalloD12 committed:
- Javadoc added.
- Removed entities files not needed for testing.

MalloD12 committed:
Function added to check whether SQL type is an enum or not, and avoid breaking build when using a Hibernate version lower than 6.3

MalloD12 committed:
Trigger CI rebuild

filipelautert committed:
Merge branch 'main' into multi-schema-support

robertericreeves commented:
+1 on dropping support for db platforms that are EOL. Not reasonable to have Liquibase support EOL versions especially when they prevent forward progress on great features like this. 

@filipelautert what do you think?
